lucille-plugins/lucille-pinecone/src/test/java/com/kmwllc/lucille/pinecone/indexer/PineconeIndexerTest.java

```diff
@@ -0,0 +1,348 @@
+      TestMessenger messenger = new TestMessenger();
+      new PineconeIndexer(configGood, messenger, "testing");
+
+      assertTrue(client.constructed().size() == 1);
+      PineconeClient constructed = client.constructed().get(0);
+      assertTrue(constructorArgs.get(constructed).get(0) instanceof PineconeClientConfig);
+
+      PineconeClientConfig config = (PineconeClientConfig) constructorArgs.get(constructed).get(0);
+
+      assertEquals("apiKey", config.getApiKey());
+      assertEquals("environment", config.getEnvironment());
+      assertEquals("projectName", config.getProjectName());
+      assertEquals(100, config.getServerSideTimeoutSec());
+    }
+  }
+
+  @Test
+  public void testValidateConnection() {
+    try (MockedConstruction<PineconeClient> client = Mockito.mockConstruction(PineconeClient.class, (mock, context) -> {
+      Mockito.when(mock.connect("good")).thenReturn(goodConnection);
+      Mockito.when(mock.connect("failure")).thenReturn(failureConnection);
+      Mockito.when(mock.connect("shutdown")).thenReturn(shutdownConnection);
+    })) {
+      TestMessenger messenger = new TestMessenger();
+      Config configGood = ConfigFactory.load("PineconeIndexerTest/good-config.conf");
+      Config configFailure = ConfigFactory.load("PineconeIndexerTest/failure-config.conf");
+      Config configShutdown = ConfigFactory.load("PineconeIndexerTest/shutdown-config.conf");
+
+      PineconeIndexer indexerGood = new PineconeIndexer(configGood, messenger, "testing");
+      PineconeIndexer indexerFailure = new PineconeIndexer(configFailure, messenger, "testing");
+      PineconeIndexer indexerShutdown = new PineconeIndexer(configShutdown, messenger, "testing");
+
+      indexerGood.validateConnection();
+      Mockito.verify(client.constructed().get(0), Mockito.times(1)).connect("good");
+      indexerGood.validateConnection();
+      Mockito.verify(client.constructed().get(0), Mockito.times(1)).connect("good");
+
+      assertTrue(indexerGood.validateConnection());
+      assertFalse(indexerFailure.validateConnection());
+      assertFalse(indexerShutdown.validateConnection());
+    }
+  }
+
+  @Test
+  public void testCloseConnection() {
+    try (MockedConstruction<PineconeClient> client = Mockito.mockConstruction(PineconeClient.class, (mock, context) -> {
+      Mockito.when(mock.connect("good")).thenReturn(goodConnection);
+    })) {
+      TestMessenger messenger = new TestMessenger();
+      Config configGood = ConfigFactory.load("PineconeIndexerTest/good-config.conf");
+
+      PineconeIndexer indexerGood = new PineconeIndexer(configGood, messenger, "testing");
+      indexerGood.closeConnection();
+      Mockito.verify(goodConnection, Mockito.times(0)).close();
+      indexerGood.validateConnection();
+      indexerGood.closeConnection();
+      Mockito.verify(goodConnection, Mockito.times(1)).close();
+    }
+  }
+
+
+  @Test
+  public void testUpsertAndUpdateNoNamespacesProvided() throws Exception {
```

Comment:
Design question: should at least one namespace be required? Should the indexer throw an exception if none are provided? If yes, we can leave this test in place for now and create a separate ticket for the new validation functionality.

Reply:
Pinecone does not require a namespace so I don't think we should force a requirement. Technically if no namespace is set the docs are sent to the default namespace "".

Reply:
While the pinecone client may send docs to the default namespace if none is provided, I believe the way the indexer is now, if no namespaces are provided in the config, then no documents are indexed. That is why I have this test.